### PR TITLE
fix: prevent SQL injection in table name handling

### LINKED ISSUE #206: SQL Injection vulnerability in PostgreSQL bootstrapper
## Description
The PostgreSQL bootstrapper has a potential SQL injection vulnerability where `table_name` is directly passed into a SQL query without proper validation or escaping.

## Location
components/bootstrappers/postgres/src/postgres.rs
`let query = format!("SELECT * FROM {table_name}");`

### PR DESCRIPTION
**Pull request by ashishmohapatra240**
# Description

Fixes a potential SQL injection vulnerability in the PostgreSQL bootstrapper where table names were directly interpolated into SQL queries without proper validation or quoting.

- Added `quote_identifier()` helper function to validate and safely quote PostgreSQL identifiers
- Validation ensures identifiers only contain alphanumeric characters and underscores

## Type of change

- This pull request fixes a bug in Drasi and has an approved issue #206 

Fixes: #206

### CODE REVIEW

**Review by ruokun-niu — APPROVED** [review]

File: components/bootstrappers/postgres/src/postgres.rs [hunk 1]
```diff
@@ -29,6 +29,37 @@ use drasi_lib::channels::SourceChangeEvent;
...
+    {
+        return Err(anyhow!(
+            "Identifier contains invalid characters: {identifier}"
+        ));
+    }
+
+    // Double-quote the identifier (PostgreSQL standard for identifiers)
+    // Also escape any internal double quotes by doubling them
+    Ok(format!("\"{}\"", identifier.replace('"', "\"\"")))
+}
```
[the remark this answers is not in the document]
Comment (ashishmohapatra240) [reply components/bootstrappers/postgres/src/postgres.rs:84]: Added test coverage for the quote_identifier function including:
- Valid simple table names (users, _temp, table_123)
- Schema-qualified names (public.users, analytics.user_events)
- Edge cases (empty strings, names starting with numbers, empty parts)
- SQL injection attempt validation (verifying malicious inputs are blocked)

File: components/bootstrappers/postgres/src/postgres.rs [hunk 2]
```diff
@@ -29,6 +29,37 @@ use drasi_lib::channels::SourceChangeEvent;
...
+        .all(|c| c.is_ascii_alphanumeric() || c == '_')
+    {
+        return Err(anyhow!(
+            "Identifier contains invalid characters: {identifier}"
+        ));
+    }
+
+    // Double-quote the identifier (PostgreSQL standard for identifiers)
+    // Also escape any internal double quotes by doubling them
+    Ok(format!("\"{}\"", identifier.replace('"', "\"\"")))
```
[the remark this answers is not in the document]
Comment (ashishmohapatra240) [reply components/bootstrappers/postgres/src/postgres.rs]: Fixed by splitting on dots and validating/quoting each part separately. The function now supports identifiers like "analytics.user_events" and produces properly quoted output: "analytics"."user_events"

[the remark this answers is not in the document]
Comment (ashishmohapatra240) [reply components/bootstrappers/postgres/src/postgres.rs]: Removed the `.replace('"', "\"\"")` operation.

[the remark this answers is not in the document]
Comment (ashishmohapatra240) [reply components/bootstrappers/postgres/src/postgres.rs]: The current implementation validates that each identifier part follows PostgreSQL's rules for unquoted identifiers (alphanumeric + underscore, starting with letter/underscore) and then quotes them.

**Review by ruokun-niu — APPROVED** [review]

File: components/bootstrappers/postgres/src/postgres.rs [hunk 3]
```diff
@@ -29,6 +29,57 @@ use drasi_lib::channels::SourceChangeEvent;
...
+    for part in parts {
+        // Validate: part cannot be empty (catches ".table", "schema.", "schema..table")
+        if part.is_empty() {
+            return Err(anyhow!(
+                "Identifier contains empty part: '{identifier}'"
+            ));
+        }
+
+        // Validate: must start with a letter or underscore
+        let first_char = part.chars().next().unwrap();
```
Comment (amansinghoriginal) [inline components/bootstrappers/postgres/src/postgres.rs]: This project enforces `clippy::unwrap_used`, so this will fail CI. Since there is already an `is_empty()` check above, you can safely use:

```rust
let Some(first_char) = part.chars().next() else {
    return Err(anyhow!("Identifier part is empty"));
};
```

Alternatively, you could combine the empty check and first char extraction into one step.

File: components/bootstrappers/postgres/src/postgres.rs [hunk 4]
```diff
@@ -29,6 +29,57 @@ use drasi_lib::channels::SourceChangeEvent;
...
+            .chars()
+            .all(|c| c.is_ascii_alphanumeric() || c == '_')
+        {
+            return Err(anyhow!(
+                "Identifier part contains invalid characters: '{part}'"
+            ));
+        }
+
+        // Double-quote the identifier part (PostgreSQL standard for identifiers)
+        quoted_parts.push(format!("\"{}\"", part));
```
Comment (amansinghoriginal) [inline components/bootstrappers/postgres/src/postgres.rs]: While your validation prevents double quotes from appearing in `part`, please add escaping here as a defense-in-depth measure:

```rust
quoted_parts.push(format!("\"{}\"", part.replace('"', "\"\"")));
```

This ensures safety even if the validation logic is modified in the future.

File: components/bootstrappers/postgres/src/postgres.rs [hunk 5]
```diff
@@ -759,3 +812,132 @@ struct ColumnInfo {
     name: String,
     type_oid: i32,
 }
+
```
Comment (amansinghoriginal) [inline components/bootstrappers/postgres/src/postgres.rs:823]: Good job on the consistent `test_quote_identifier_*` naming pattern across all test functions. This makes the test output easy to read and filter.

**Review by amansinghoriginal — CHANGES_REQUESTED** [review]
Thank you for working on this fix. The overall approach of validation plus identifier quoting is correct, and the test coverage is thorough.

There are a few changes needed before this can be merged:

1. Replace `.unwrap()` with proper error handling (line 59) - this will fail CI otherwise
2. Add double-quote escaping as defense-in-depth (line 77)
3. Add validation in `map_labels_to_tables()` to catch invalid labels earlier in the pipeline

**Regarding point 3:** Please also add validation in `map_labels_to_tables()` (around line 429) where labels are first converted to table names. This catches invalid input earlier with clearer error context:

```rust
for label in all_labels {
    let table_name = label.to_lowercase();
    
    // Validate identifier characters early
    if !table_name.chars().all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '.') {
        return Err(anyhow!("Label '{}' contains invalid characters for table name", label));
    }
    
    // ... rest of existing logic
}
```

This follows the fail-fast principle and provides users with more actionable error messages.

See inline comments for additional details. Once these are addressed, this will be ready to merge.

File: components/bootstrappers/postgres/src/postgres.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/bootstrappers/postgres/src/postgres.rs]: I noticed Copilot suggested removing the escaping since validation prevents double quotes from appearing. While technically correct, I recommend adding it back as a defense-in-depth measure.

Security-critical code benefits from layered defenses. If the validation logic is ever relaxed in the future (e.g., to support more PostgreSQL identifier characters), the escaping ensures the quoting remains safe. The runtime cost is negligible - it is effectively a no-op when no double quotes exist.

This is a common pattern in security code: validate strictly AND sanitize output, even when one makes the other redundant today.

File: components/bootstrappers/postgres/src/postgres.rs (hunk 3 above)
Comment (ashishmohapatra240) [reply components/bootstrappers/postgres/src/postgres.rs]: Done 👍

File: components/bootstrappers/postgres/src/postgres.rs [hunk 6]
```diff
@@ -29,6 +29,60 @@ use drasi_lib::channels::SourceChangeEvent;
...
+/// Validates and quotes a PostgreSQL identifier to prevent SQL injection.
+///
+/// Supports schema-qualified identifiers (e.g., "schema.table") by splitting on dots
+/// and validating/quoting each part individually.
+///
+/// Each part must:
+/// - Not be empty
+/// - Start with a letter or underscore
+/// - Contain only alphanumeric characters and underscores
+fn quote_identifier(identifier: &str) -> Result<String> {
```
Comment (danielgerlag) [inline components/bootstrappers/postgres/src/postgres.rs:41]: Would the following not suffice?:

```rust
fn quote_identifier(s: &str) -> String {
    format!("\"{}\"", s.replace('"', "\"\""))
}
```
